pkg/api/norman/customization/cluster/validator.go

```diff
@@ -332,6 +336,162 @@ func (v *Validator) validateGenericEngineConfig(request *types.APIContext, spec
+	if err != nil {
+		return httperror.NewAPIError(httperror.ServerError, fmt.Sprintf("failed to confirm displayName is unique among Rancher AKS clusters for location %s", location))
+	}
+
+	for _, cluster := range clusters.Items {
+		if cluster.Spec.AKSConfig == nil {
+			continue
+		}
+		if name != cluster.Spec.AKSConfig.DisplayName {
+			continue
+		}
+		if location != cluster.Spec.AKSConfig.ResourceLocation {
+			continue
+		}
+		return httperror.NewAPIError(httperror.InvalidBodyContent, fmt.Sprintf("cluster already exists for AKS cluster [%s] in location [%s]", name, location))
+	}
+
+	return nil
+}
+
+// validateAKSCredentialAuth validates that a user has access to the credential they are setting and the credential
+// they are overwriting. If there is no previous credential such as during a create or the old credential cannot
+// be found, the auth check will succeed as long as the user can access the new credential.
+func validateAKSCredentialAuth(request *types.APIContext, credential string, prevCluster *v3.Cluster) error {
+	var accessCred mgmtclient.CloudCredential
+	credentialErr := "error accessing cloud credential"
+	if err := access.ByID(request, &mgmtSchema.Version, mgmtclient.CloudCredentialType, credential, &accessCred); err != nil {
+		return httperror.NewAPIError(httperror.NotFound, credentialErr)
+	}
+
+	if prevCluster == nil {
+		return nil
+	}
+
+	if prevCluster.Spec.AKSConfig == nil {
+		return nil
+	}
+
+	// validate the user has access to the old cloud credential before allowing them to change it
+	credential = prevCluster.Spec.AKSConfig.AzureCredentialSecret
+	if err := access.ByID(request, &mgmtSchema.Version, mgmtclient.CloudCredentialType, credential, &accessCred); err != nil {
+		if apiError, ok := err.(*httperror.APIError); ok {
+			if apiError.Code.Status == httperror.NotFound.Status {
+				// old cloud credential doesn't exist anymore, anyone can change it
+				return nil
+			}
+		}
+		return httperror.NewAPIError(httperror.NotFound, credentialErr)
+	}
+
+	return nil
+}
+
+// validateAKSKubernetesVersion checks whether a kubernetes version is provided
+func validateAKSKubernetesVersion(spec *v32.ClusterSpec, prevCluster *v3.Cluster) error {
+	clusterVersion := spec.AKSConfig.KubernetesVersion
+
+	if clusterVersion == nil {
+		return httperror.NewAPIError(httperror.InvalidBodyContent, "cluster kubernetes version cannot be empty string")
+	}
+
+	return nil
+}
+
+// validateAKSNodePools checks whether a given NodePool version is empty or not supported.
+// More involved validation is performed in the aks-operator.
+func validateAKSNodePools(spec *v32.ClusterSpec) error {
+	nodePools := spec.AKSConfig.NodePools
+	if nodePools == nil {
+		return nil
+	}
+	if len(nodePools) == 0 {
+		return httperror.NewAPIError(httperror.InvalidBodyContent, fmt.Sprintf("must have at least one nodepool"))
+	}
+
+	var errors []string
+
+	for _, np := range nodePools {
+		name := np.Name
+		if name == nil {
```

Comment:
Should we also check that the value referenced by `np.Name` is not the empty string?

Reply:
Checked